File: 2024/Day 3: Mull It Over/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	memory, _ := readMemory()
	calculateOnlyActives(memory)
	calculateAll(memory)
}

func calculateOnlyActives(memory string) {
	activeCommands := getActiveCommands(memory)
	result := 0
	for _, activeCommand := range activeCommands {
		mulInput := getMulCommands(activeCommand)
		result += calculate(mulInput)
	}
	fmt.Println("active calculation: " + strconv.Itoa(result))
}

func calculateAll(memory string) {
	mulInput := getMulCommands(memory)
	fmt.Println("all calculation: " + strconv.Itoa(calculate(mulInput)))
}

func calculate(mulInput []string) int {
	result := 0
	for _, input := range mulInput {
		mulResult, err := mul(input)
		if err != nil {
			fmt.Println(err.Error())
			os.Exit(1)
		}
		result += mulResult
	}
	return result
}

func getMulCommands(memory string) []string {
	startingInput := strings.Split(memory, "mul(")
	mulInput := make([]string, 0)
	for _, start := range startingInput {
		if strings.Contains(start, ")") {
			endInput := strings.Split(start, ")")
			match, _ := regexp.MatchString("^\\d*,\\d*$", endInput[0])
			if match {
				mulInput = append(mulInput, endInput[0])
			}
		}
	}
	return mulInput
}

func getActiveCommands(memory string) []string {
	startDoes := strings.Split(memory, "do()")
	does := make([]string, len(startDoes))
	for i, startDo := range startDoes {
		does[i] = strings.Split(startDo, "don't")[0]
	}
	return does
}

func mul(input string) (int, error) {
	inputSplit := strings.Split(input, ",")
	a, err := strconv.Atoi(inputSplit[0])
	if err != nil {
		return 0, err
	}
	b, err := strconv.Atoi(inputSplit[1])
	if err != nil {
		return 0, err
	}
	return a * b, nil
}

func readMemory() (string, error) {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("Enter input (press Ctrl+D or Ctrl+Z to end):")
	memory := ""
	for scanner.Scan() {
		reportLine := scanner.Text()
		memory += reportLine
	}

	if err := scanner.Err(); err != nil {
		return "", err
	}
	return memory, nil
}

File: 2024/Day 3: Mull It Over/java/Main.java
import java.util.*;
import java.util.regex.*;
import java.util.stream.Collectors;

public class Main {

    public static void main(String[] args) {
        try {
            String memory = readMemory();
            calculateOnlyActives(memory);
            calculateAll(memory);
        } catch (RuntimeException e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }

    private static void calculateOnlyActives(String memory) {
        var result = getActiveCommands(memory).stream()
                .map(Main::getMulCommands)
                .map(Main::calculate)
                .filter(Optional::isPresent)
                .map(Optional::get)
                .reduce(Integer::sum);
        System.out.println("active calculation: " + result.orElse(0));
    }

    private static void calculateAll(String memory) {
        List<String> mulInput = getMulCommands(memory);
        System.out.println("all calculation: " + calculate(mulInput).orElse(0));
    }

    private static Optional<Integer> calculate(List<String> mulInput) {
        return mulInput.stream().map(Main::mul).reduce(Integer::sum);
    }

    private static List<String> getMulCommands(String memory) {
        String[] startingInput = memory.split("mul\\(");
        return Arrays.stream(startingInput)
                .filter(start -> start.contains(")"))
                .map(start -> start.split("\\)"))
                .filter(endInput -> Pattern.matches("^\\d*,\\d*$", endInput[0]))
                .map(endInput -> endInput[0])
                .toList();
    }

    private static List<String> getActiveCommands(String memory) {
        String[] startDoes = memory.split("do\\(\\)");
        return Arrays.stream(startDoes)
                .map(startDo -> startDo.split("don't")[0])
                .collect(Collectors.toList());
    }

    private static int mul(String input)  {
        String[] inputSplit = input.split(",");
        if (inputSplit.length != 2) {
            throw new IllegalArgumentException("Invalid mul input: " + input);
        }
        int a = Integer.parseInt(inputSplit[0]);
        int b = Integer.parseInt(inputSplit[1]);
        return a * b;
    }

    private static String readMemory() {
        StringBuilder memory = new StringBuilder();
        try (Scanner scanner = new Scanner(System.in)) {
            System.out.println("Enter input (press Ctrl+D or Ctrl+Z to end):");
            while (scanner.hasNextLine()) {
                memory.append(scanner.nextLine());
            }
        }
        return memory.toString();
    }
}
